controllers/fluxcd/application-controller.go

```diff
@@ -0,0 +1,453 @@
+	r.recorder.Eventf(kus, corev1.EventTypeNormal, "Created", "Created FluxCD Kustomization %s", kus.GetAnnotations()["app.kubernetes.io/name"])
+	return
+}
+func (r *ApplicationReconciler) updateKustomization(ctx context.Context, app *v1alpha1.Application, kus *kusv1.Kustomization, deploy *v1alpha1.KustomizationSpec) (err error) {
+	newKus := buildKustomization(app, deploy)
+	kus.Spec = newKus.Spec
+	if err = r.Update(ctx, kus); err != nil {
+		return
+	}
+	r.log.Info("Update FluxCD HelmRelease", "name", kus.GetAnnotations()["app.kubernetes.io/name"])
+	r.recorder.Eventf(kus, corev1.EventTypeNormal, "Updated", "Update FluxCD HelmRelease %s", kus.GetAnnotations()["app.kubernetes.io/name"])
+	return
+}
+
+func isHelmOrKustomize(app *v1alpha1.Application) (AppType, error) {
+	conf := app.Spec.FluxApp.Spec.Config.DeepCopy()
+	helm, kus := conf.HelmRelease != nil, conf.Kustomization != nil
+	if helm && !kus {
+		return HelmRelease, nil
+	}
+	if !helm && kus {
+		return Kustomization, nil
+	}
+	return "", fmt.Errorf("should has one and only one FluxApplicationConfig (HelmRelease or Kustomization)")
+}
+
+// GetGroupName returns the group name
+func (r *ApplicationReconciler) GetGroupName() string {
+	return controllerGroupName
+}
+
+// GetName returns the name of this reconciler
+func (r *ApplicationReconciler) GetName() string {
+	return "FluxApplicationReconciler"
+}
+
+func getHelmReleaseName(deploy *v1alpha1.Deploy) string {
+	// host cluster
+	if deploy.Destination.KubeConfig == nil {
+		return deploy.Destination.TargetNamespace
+	}
+	// member cluster
+	return deploy.Destination.KubeConfig.SecretRef.Name + "-" + deploy.Destination.TargetNamespace
+}
+
+func getKustomizationName(deploy *v1alpha1.KustomizationSpec) string {
+	// host cluster
+	if deploy.Destination.KubeConfig == nil {
+		return deploy.Destination.TargetNamespace
+	}
+	// member cluster
+	return deploy.Destination.KubeConfig.SecretRef.Name + "-" + deploy.Destination.TargetNamespace
+}
+
+func (r *ApplicationReconciler) saveTemplate(ctx context.Context, helmChart *sourcev1.HelmChart) (err error) {
+	ns, name := helmChart.GetNamespace(), helmChart.GetName()
+
+	if err = r.Get(ctx, types.NamespacedName{Namespace: ns, Name: name}, helmChart); err != nil {
+		if !apierrors.IsNotFound(err) {
+			return
+		}
+		// not found
+		// create
+		if err = r.Create(ctx, helmChart); err != nil {
+			return
+		}
+		r.log.Info("Create HelmTemplate", "name", helmChart.GetName())
+		r.recorder.Eventf(helmChart, corev1.EventTypeNormal, "Created", "Created HelmTemplate %s", helmChart.GetName())
+	}
+	return
+}
+
+func wantSaveHelmTemplate(app *v1alpha1.Application) bool {
+	if v, ok := app.GetLabels()[v1alpha1.SaveTemplateLabelKey]; ok {
+		return v == "true"
+	}
+	return false
+}
+
+func buildTemplateFromApp(app *v1alpha1.Application) *sourcev1.HelmChart {
+	s := app.Spec.FluxApp.Spec.Source.SourceRef.DeepCopy()
```

Comment:
`app.Spec.FluxApp.Spec.Source` is a pointer. It would be nil. It's not safe to use those values directly. I know you already checked if app.Spec.FluxApp is nil somewhere. But it's still not safe if other contributors try to use this function without going through all the code lines.

You could consider only passing `app.Spec.FluxApp` directly. Because it's not necessary to know all the structure.

Reply:
Thanks. Really appreciate your advice.